src/Mvc/Mvc.ViewFeatures/src/Rendering/HtmlHelperComponentExtensions.cs

```diff
@@ -58,7 +59,21 @@ public static class HtmlHelperComponentExtensions
             ParameterView.FromDictionary(HtmlHelper.ObjectToDictionary(parameters));
 
         var httpContext = htmlHelper.ViewContext.HttpContext;
-        var componentRenderer = httpContext.RequestServices.GetRequiredService<ComponentPrerenderer>();
-        return await componentRenderer.PrerenderComponentAsync(httpContext, componentType, renderMode, parameterView);
+        var componentRenderer = httpContext.RequestServices.GetRequiredService<IComponentPrerenderer>();
+        return await componentRenderer.PrerenderComponentAsync(httpContext, componentType, MapRenderMode(renderMode), parameterView);
     }
+
+    // This is unfortunate, and we might want to find a better solution. There's an existing public
+    // type Microsoft.AspNetCore.Mvc.Rendering.RenderMode which we now need to use in a lower layer,
+    // M.A.C.Endpoints. Even type-forwarding is not a good solution because we really want to change
+    // the namespace. So this code maps the old enum to the newer one.
```

Comment:
We should consider taking a breaking change for this. We could make it usually *not* source-breaking (although it would always be binary-breaking).

That is, if we entirely deleted the old enum `Microsoft.AspNetCore.Mvc.Rendering.RenderMode` and then were only left with the new `Microsoft.AspNetCore.Components.RenderMode` one - and all the enum values are the same - then in most cases people's existing source code would compile unchanged, and in the worst case they would just have to add `using Microsoft.AspNetCore.Components;`.

The alternatives are:

 * Do nothing - and then we have to live with the ambiguity between `Microsoft.AspNetCore.Mvc.Rendering.RenderMode` and `Microsoft.AspNetCore.Components.RenderMode`
 * Mark `Microsoft.AspNetCore.Mvc.Rendering.RenderMode` as obsolete (and maybe `EditorVisible.Never` to reduce any cases where the IDE prompts to add the namespace `Microsoft.AspNetCore.Mvc.Rendering` for this), and add an extra overload to `Html.RenderComponentAsync` to take the new enum, and unfortunately add a completely separate property on `ComponentTagHelper` to live in parallel with the existing `public RenderMode RenderMode { get; set; }` one. Not sure if there are any better ways for a tag helper attribute to change its type.

Reply:
I would be ok I think if we do this breaking change. We are pretty early in the release, and I suspect not many people will be affected.

The biggest issue would be if there is a library out somewhere that ships a package that uses this type, but I don't think that will be super common.

Since the scope of what we are doing is big, I'd rather make sure we get the layering right and cause pain once than having to live with it for eternity, especially for such a small thing like this.

Reply:
OK, filed https://github.com/dotnet/aspnetcore/issues/47133